src/accounts.rs

```diff
@@ -31,84 +35,386 @@ pub struct ErrorCounters {
+            let result: Option<u64>;
+            {
+                if account.tokens != 0 {
+                    acc = account;
+                }
+                let av = &self.storage.read().unwrap()[id].appendvec;
+                result = av.read().unwrap().append_account(&acc);
+            }
+            if let Some(val) = result {
+                offset = val;
+                break;
+            } else {
+                id = self.get_storage_id(start, id);
+            }
+        }
+        (id, offset)
+    }
+
     /// Store the account update.  If the update is to delete the account because the token balance
     /// is 0, purge needs to be set to true for the delete to occur in place.
-    pub fn store(&mut self, purge: bool, pubkey: &Pubkey, account: &Account) {
-        if account.tokens == 0 {
-            if purge {
-                // purge if balance is 0 and no checkpoints
-                self.accounts.remove(pubkey);
-            } else {
-                // store default account if balance is 0 and there's a checkpoint
-                self.accounts.insert(pubkey.clone(), Account::default());
+    pub fn store(&self, fork: Fork, purge: bool, pubkey: &Pubkey, account: &Account) {
+        if account.tokens == 0 && purge {
+            // purge if balance is 0 and no checkpoints
+            let mut index = self.index_info.index.write().unwrap();
+            if let Some(forks) = index.remove(pubkey) {
+                let stores = self.storage.read().unwrap();
+                for (_, (id, _)) in forks.0.iter() {
+                    stores[*id].count.fetch_sub(1, Ordering::Relaxed);
+                }
+            }
+            if vote_program::check_id(&account.owner) {
+                self.index_info.vote_index.write().unwrap().remove(pubkey);
             }
         } else {
-            self.accounts.insert(pubkey.clone(), account.clone());
+            let (id, offset) = self.append_account(&account);
+
+            if vote_program::check_id(&account.owner) {
+                let mut index = self.index_info.vote_index.write().unwrap();
+                if account.tokens == 0 {
+                    index.remove(pubkey);
+                } else {
+                    index.insert(*pubkey);
+                }
+            }
+
+            let mut result: Option<usize> = None;
+            {
+                let mut insert: Option<usize> = None;
+                let mut windex = self.index_info.index.write().unwrap();
+                let forks = windex.entry(*pubkey).or_insert(AccountMap(vec![]));
+                for (i, (v_fork, (v_id, _))) in forks.0.iter().enumerate() {
+                    if *v_fork > fork {
+                        continue;
+                    }
+                    if *v_fork == fork {
+                        result = Some(*v_id);
+                        forks.0[i] = (fork, (id, offset));
+                        break;
+                    }
+                    insert = Some(i);
```

Comment:
While storing the index, it adds it to the Vec as a sorted fashion (based on bank tick height). Would this not work?

Reply:
Is that because the bank doesn’t know it’s own fork id?

Reply:
when you say fork id, I am interpreting it as that of bank id used in BankForks. These are inserted in sorted order into the accounts index, This way we could bail out early rather than have to iterate the entire list everytime.